server/src/main/java/org/elasticsearch/rest/action/document/RestBulkAction.java

```diff
@@ -66,38 +76,124 @@ public String getName() {
 
     @Override
     public RestChannelConsumer prepareRequest(final RestRequest request, final NodeClient client) throws IOException {
+        // TODO: Move this to CTOR and hook everything up
+        if (bulkHandler == null) {
+            bulkHandler = new IncrementalBulkHandler(client);
+        }
+
         if (request.getRestApiVersion() == RestApiVersion.V_7 && request.hasParam("type")) {
             request.param("type");
         }
-        BulkRequest bulkRequest = new BulkRequest();
-        String defaultIndex = request.param("index");
-        String defaultRouting = request.param("routing");
-        FetchSourceContext defaultFetchSourceContext = FetchSourceContext.parseFromRestRequest(request);
-        String defaultPipeline = request.param("pipeline");
-        boolean defaultListExecutedPipelines = request.paramAsBoolean("list_executed_pipelines", false);
-        String waitForActiveShards = request.param("wait_for_active_shards");
-        if (waitForActiveShards != null) {
-            bulkRequest.waitForActiveShards(ActiveShardCount.parseString(waitForActiveShards));
+
+        return new ChunkHandler(request);
+    }
+
+    private class ChunkHandler implements BaseRestHandler.RequestBodyChunkConsumer {
+
+        private final RestRequest request;
+
+        private final Map<String, String> stringDeduplicator = new HashMap<>();
+        private final String defaultIndex;
+        private final String defaultRouting;
+        private final FetchSourceContext defaultFetchSourceContext;
+        private final String defaultPipeline;
+        private final boolean defaultListExecutedPipelines;
+        private final Boolean defaultRequireAlias;
+        private final boolean defaultRequireDataStream;
+        private final BulkRequestParser parser;
+        private final IncrementalBulkHandler.Handler handler;
+
+        private volatile RestChannel restChannel;
+        private final ArrayDeque<ReleasableBytesReference> unParsedChunks = new ArrayDeque<>(4);
+        private final ArrayList<DocWriteRequest<?>> items = new ArrayList<>(4);
```

Comment:
I see reuse of `items`, but cannot find where list is cleared after passing to `handler.addItems`